[ENH] `skforecast` integration for time series hyperparameter tuning

Omswastik-11 opened the pull request:
## Summary

This PR adds a full integration with **skforecast**, allowing Hyperactive to optimize hyperparameters of skforecast forecasting models using any of its optimization algorithms.

## Implementation Details

### SkforecastExperiment (skforecast_forecasting.py)

* Inherits from `BaseExperiment`.
* Uses `skforecast.model_selection.backtesting_forecaster` inside `_evaluate()` to perform time-series cross-validation for each parameter set.
* clones the forecaster and applies new parameters with `set_params()` before every evaluation.

### SkforecastOptCV (skforecast_opt_cv.py)

* sklearn-style estimator (inherits from `BaseEstimator`).
* Works with `ForecasterRecursive` and other compatible skforecast forecasters.
* **fit():**

  * Builds a `SkforecastExperiment` with user settings (`steps`, `initial_train_size`, `metric`, etc.).
  * runs Hyperactive’s optimizer to search for the best hyperparameters.
  * refits the best forecaster on all available data.
* **predict():**

  * delegates to `best_forecaster_.predict()` for easy forecasting after optimization.

## Configuration

* added **skforecast** as an optional dependency in `pyproject.toml` under the `integrations` extra.

## Verification

* added `skforecast_example.py` showing a HillClimbing search with `ForecasterRecursive` + `RandomForestRegressor`.
* added unit tests to verify parameter handling, experiment execution, and integration flow.

## Closes

**Fixes #199**

fkiraly commented on src/hyperactive/experiment/integrations/skforecast_forecasting.py:
```diff
@@ -0,0 +1,146 @@
+    initial_train_size : int
+        Number of samples in the initial training set
+
+    exog : pandas Series or DataFrame, optional
```
I would move it up to `y` directly.

fkiraly reviewed (changes requested):
Thanks, great!

* please improve docstrings, see above, and also include defaults
* please add `get_test_params` with sensible settings to test the experiment and the forecaster

Omswastik-11 commented:
Hi @fkiraly !! 
Modified the docs string. As you suggested .

added `get_test_params()` verified all the tests . 

tested the pre-commit on changed files 
<img width="1437" height="247" alt="image" src="https://github.com/user-attachments/assets/0b2ac10b-d880-4e06-805f-6284303f1110" />

kindly verify this.

fkiraly reviewed (changes requested):
For the tests to run, you need to add `skforecast` to the python environment - I would add it to `sktime-integration` in `pyproject.toml`, that might be easiest.

JavierEscobarOrtiz commented on src/hyperactive/integrations/skforecast/skforecast_opt_cv.py:
```diff
@@ -0,0 +1,200 @@
+
+        return self
+
+    def predict(self, steps, exog=None, **kwargs):
```
I will also include the `last_window` parameter in the `predict` method. You can copy the docstring from here:

https://skforecast.org/0.18.0/api/forecasterrecursive#skforecast.recursive._forecaster_recursive.ForecasterRecursive.predict

`check_inputs` is for internal use only, so it is not worth including it.

JoaquinAmatRodrigo commented:
It needs to be triggered by one of the repository's maintainers.

Omswastik-11 commented:
Hi @JoaquinAmatRodrigo, @SimonBlanke, and @fkiraly — I’d appreciate your suggestions on how to proceed here.

The issue is that **`skforecast` currently does not support Python 3.14**, which affects the new integration.
I am a bit unsure about two things:

1. **Where exactly should `skforecast` be declared?**
   – Only in `pyproject.toml` under optional extras?
   – Or in the  `_tags`?
2. **When should `test:vm` be enabled or disabled?**
---

## What I have done so far

### **1. Fixes for the `test_examples` workflow**

Example test run:
[https://github.com/SimonBlanke/Hyperactive/actions/runs/19608701548/job/56518055022?pr=208](https://github.com/SimonBlanke/Hyperactive/actions/runs/19608701548/job/56518055022?pr=208)

I checked the `examples/` directory and noticed that many dependencies (e.g., `torch`, `tensorflow`) were not actually needed.

#### Changes:

**`pyproject.toml`**

* Added a new `test_examples` optional dependency group including:

  * `pytest`
  * integrations used in examples (`sklearn`, `sktime`, `skforecast`)
  * `optuna`

**`Makefile`**

* Added a new target for running example tests.

**`test.yml`**

* Updated the `test-examples` job to install only the required libraries .
---

### **2. Handling the Python 3.14 compatibility issue**

`skforecast` fails to install on Python 3.14, so I added conditional version constraints:

```toml
skforecast; python_version < "3.14"
```

This has been added to all optional dependency groups that require skforecast.

Related CI runs:
[https://github.com/SimonBlanke/Hyperactive/actions/runs/19608701548/job/56518055064?pr=208](https://github.com/SimonBlanke/Hyperactive/actions/runs/19608701548/job/56518055064?pr=208)
[https://github.com/SimonBlanke/Hyperactive/actions/runs/19608701548/job/56518055077?pr=208](https://github.com/SimonBlanke/Hyperactive/actions/runs/19608701548/job/56518055077?pr=208)

---

### local testing

<img width="1092" height="218" alt="image" src="https://github.com/user-attachments/assets/5f27a296-136d-4d88-9d9a-c44eaa470c9b" />


Thanks!

fkiraly commented on src/hyperactive/experiment/integrations/skforecast_forecasting.py:
```diff
@@ -0,0 +1,207 @@
+            instance.
+            create_test_instance uses the first (or only) dictionary in `params`
+        """
+        import numpy as np
```
this is causing test collection issues - I would use `_check_soft_dependencies` from `skbase.utils.dependencies` to fill `None` in cased `skforecast` is not present, i.e., return a `dict` that fills `None` for the required params, to avoid a failing import.

JoaquinAmatRodrigo commented:
> The integration works, nice!
> 
> The remaining issues relate to dependency isolation, you have to do it in two places:
> 
> * the `get_test_params` function, see above
> * in the tags (I think you also need to add a few others)
> 
> Besides this, the "higher is better" property needs to be inferred from the `metric` and set as a tag in `__init__`.
> 
> @JoaquinAmatRodrigo, is there a programmatic way to do this?

We do not have a programmatic strategy for this. So far, all the regression metrics that Skforecast allows to be passed as a string are intended to be minimised.

```python
"mean_squared_error": mean_squared_error,
"mean_absolute_error": mean_absolute_error,
"mean_absolute_percentage_error": mean_absolute_percentage_error,
"mean_squared_log_error": mean_squared_log_error,
"mean_absolute_scaled_error": mean_absolute_scaled_error,
"root_mean_squared_scaled_error": root_mean_squared_scaled_error,
"median_absolute_error": median_absolute_error,
"symmetric_mean_absolute_percentage_error": symmetric_mean_absolute_percentage_error
```
If the user passes a custom function as a metric, they need to indicate whether it is a maximisation or minimisation.

JoaquinAmatRodrigo commented:
Looks like a great solution. You might want to double-check the keywords, usually libraries use 'maximize' or 'minimize'. Not  sure what is the ones used in this library.

Omswastik-11 commented:
Thanks @JoaquinAmatRodrigo !! 
In my recent changes I have added a boolean parameter named 'higher_is_better' . 
https://github.com/SimonBlanke/Hyperactive/pull/208/commits/ca013a4c83ce583a8fa0766dd6416cb7e2953e5c

I checked `sktime_forecasting.py` they used scoring as parameter for both metrics and higher_or_lower part but as we have one separate parameter for metric so i just added a boolean param .
https://github.com/SimonBlanke/Hyperactive/blob/main/src/hyperactive/experiment/integrations/sktime_forecasting.py#L171

@fkiraly and @SimonBlanke any suggestions ?

Omswastik-11 commented:
Hi @fkiraly !!! 
Have a look at this . I explained why I did changes
https://github.com/SimonBlanke/Hyperactive/pull/208#issuecomment-3585153351

fkiraly commented:
> Have a look at this . I explained why I did changes

What you say in the comment you mention is not consistent with the actual changes in the `pyproject`.

If you are using AI, please watch what it is doing.

Omswastik-11 commented:
Hi @SimonBlanke !! Can you re run the workflow ? I have reverted the changes and added this which last time solved the storage issue .

refer this 
https://github.com/hyperactive-project/Hyperactive/actions/runs/19744054270/job/56593197910?pr=208
```
  - name: Free Disk Space (Ubuntu)
    if: runner.os == 'Linux'
    run: |
      sudo rm -rf /usr/share/dotnet
      sudo rm -rf /usr/local/lib/android
      sudo rm -rf /opt/ghc
      sudo rm -rf /opt/hostedtoolcache/CodeQL
      sudo docker image prune --all --force
```

Omswastik-11 commented on .github/workflows/test.yml:
```diff
@@ -108,6 +108,15 @@ jobs:
+          sudo rm -rf /opt/ghc
+          sudo rm -rf /opt/hostedtoolcache/CodeQL
+          sudo docker image prune --all --force
+
```
Thanks @SimonBlanke !! Can you suggest what alternate solution could work ?

Is it ok ? If we just use these commands ?

```
  - name: Free Disk Space (Ubuntu)
    if: runner.os == 'Linux'
    run: |
      sudo rm -rf /usr/share/dotnet
      sudo rm -rf /usr/local/lib/android
```
